python/pyspark/sql/connect/session.py

```diff
@@ -385,6 +401,120 @@ def udf(self) -> Any:
+          Spark Connect are not bundled in the regular Apache Spark release).
+
+        2. Temporarily remove all states for Spark Connect, for example, ``SPARK_REMOTE``
+          environment variable.
+
+            2.1. If we're in PySpark application submission, e.g., ``bin/spark-submit app.py``
+              starts a JVM (without Spark Context) first, and adds the Spark Connect server jars
+              into the current class loader. Otherwise, Spark Context with ``spark.plugins``
+              cannot be initialized because the JVM is already running without the jars in
+              the class path before executing this Python process for driver side.
+
+        3. Starts a regular Spark session that automatically starts a Spark Connect server
+          with JVM (if it is not up) via ``spark.plugins`` feature.
+
+        The second case is when you use Apache Spark release:
+
+        1. Users must specify either the jars or package, e.g., ``--packages
+          org.apache.spark:spark-connect_2.12:3.4.0``. The jars or packages would be specified
+          in SparkSubmit automatically. This method does not do anything related to this.
+
+        2. Temporarily remove all states for Spark Connect, for example, ``SPARK_REMOTE``
+          environment variable. It does not do anything for PySpark application submission as
+          well because jars or packages were already specified before executing this Python
+          process for driver side.
+
+        3. Starts a regular Spark session that automatically starts a Spark Connect server
+          with JVM via ``spark.plugins`` feature.
+        """
+        session = PySparkSession._instantiatedSession
+        if session is None or session._sc._jsc is None:
+            conf = SparkConf()
+            # Check if we're using unreleased version that is in development.
+            is_dev_version = "dev" in LooseVersion(__version__).version
+            connect_jar = None
+
+            if is_dev_version:
+                from pyspark.testing.utils import search_jar
+
+                # Note that, in production, spark.jars.packages configuration should be
+                # set by users. Here we're automatically searching the jars locally built.
+                connect_jar = search_jar(
+                    "connector/connect/server", "spark-connect-assembly-", "spark-connect"
+                )
+                if connect_jar is None:
+                    raise RuntimeError(
+                        "Spark Connect project was not compiled into a JAR. "
+                        "To use --remote without specifying jars, you need to build Spark "
+                        "with 'build/sbt package' or 'build/mvn package' first"
+                    )
+                conf.set("spark.jars", connect_jar)
+
+            conf.set("spark.master", master)
+            conf.set("spark.plugins", "org.apache.spark.sql.connect.SparkConnectPlugin")
```

Comment:
nit, maybe append  `connect_jar` into `spark.jars`, and `SparkConnectPlugin` into `spark.plugins`, rather than overwrite them?

Reply:
Good point!